File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/drive/opmode/AutomaticFeedforwardTuner.java
package org.firstinspires.ftc.teamcode.drive.opmode;

import static org.firstinspires.ftc.teamcode.drive.DriveConstants.MAX_RPM;
import static org.firstinspires.ftc.teamcode.drive.DriveConstants.RUN_USING_ENCODER;
import static org.firstinspires.ftc.teamcode.drive.DriveConstants.rpmToVelocity;

import com.acmerobotics.dashboard.FtcDashboard;
import com.acmerobotics.dashboard.config.Config;
import com.acmerobotics.dashboard.telemetry.MultipleTelemetry;
import com.acmerobotics.roadrunner.geometry.Pose2d;
import com.acmerobotics.roadrunner.util.NanoClock;
import com.arcrobotics.ftclib.command.CommandOpMode;
import com.arcrobotics.ftclib.command.FunctionalCommand;
import com.arcrobotics.ftclib.command.InstantCommand;
import com.arcrobotics.ftclib.command.button.Button;
import com.arcrobotics.ftclib.command.button.GamepadButton;
import com.arcrobotics.ftclib.gamepad.GamepadEx;
import com.arcrobotics.ftclib.gamepad.GamepadKeys;
import com.qualcomm.robotcore.eventloop.opmode.Autonomous;
import com.qualcomm.robotcore.util.RobotLog;

import org.firstinspires.ftc.robotcore.internal.system.Misc;
import org.firstinspires.ftc.teamcode.drive.SampleMecanumDrive;
import org.firstinspires.ftc.teamcode.subsystems.drive.roadrunner.MecanumDriveSubsystem;
import org.firstinspires.ftc.teamcode.util.LoggingUtil;
import org.firstinspires.ftc.teamcode.util.RegressionUtil;

import java.util.ArrayList;
import java.util.List;

/**
 * Op mode for computing kV, kStatic, and kA from various drive routines. For the curious, here's an
 * outline of the procedure:
 *   1. Slowly ramp the motor power and record encoder values along the way.
 *   2. Run a linear regression on the encoder velocity vs. motor power plot to obtain a slope (kV)
 *      and an optional intercept (kStatic).
 *   3. Accelerate the robot (apply constant power) and record the encoder counts.
 *   4. Adjust the encoder data based on the velocity tuning data and find kA with another linear
 *      regression.
 *
 * NOTE: this has been refactored to use FTCLib's command-based
 */
@Config
@Autonomous(group = "drive")
public class AutomaticFeedforwardTuner extends CommandOpMode {

    public static final double MAX_POWER = 0.7;
    public static final double DISTANCE = 100; // in

    private MecanumDriveSubsystem drive;
    private GamepadEx gamepad;
    private Button aButton, bButton;
    private NanoClock clock;

    private boolean fitIntercept = false;

    private double maxVel, startTime, finalVel, rampTime, accel, elapsedTime, maxPowerTime;
    private List<Double> timeSamples, positionSamples, powerSamples;
    private RegressionUtil.RampResult rampResult;

    private State state;

    enum State {
        STATIC, ACCEL, ACCEL_STARTED
    }

    @Override
    public void initialize() {
        if (RUN_USING_ENCODER) {
            RobotLog.setGlobalErrorMsg("Feedforward constants usually don't need to be tuned " +
                    "when using the built-in drive motor velocity PID.");
        }

        telemetry = new MultipleTelemetry(telemetry, FtcDashboard.getInstance().getTelemetry());

        drive = new MecanumDriveSubsystem(new SampleMecanumDrive(hardwareMap), false);
        gamepad = new GamepadEx(gamepad1);
        state = State.STATIC;

        telemetry.addLine("Press play to begin the feedforward tuning routine");
        telemetry.update();

        schedule(new InstantCommand(() -> {
            telemetry.clearAll();
            telemetry.addLine("Would you like to fit kStatic?");
            telemetry.addLine("Press (A) for yes, (B) for no");
            telemetry.update();
        }));

        InstantCommand staticFinishedCommand = new InstantCommand(() -> {
            drive.setDrivePower(new Pose2d(0.0, 0.0, 0.0));

            rampResult = RegressionUtil.fitRampData(
                    timeSamples, positionSamples, powerSamples, fitIntercept,
                    LoggingUtil.getLogFile(Misc.formatInvariant(
                            "DriveRampRegression-%d.csv", System.currentTimeMillis())));

            telemetry.clearAll();
            telemetry.addLine("Quasi-static ramp up test complete");
            if (fitIntercept) {
                telemetry.addLine(Misc.formatInvariant("kV = %.5f, kStatic = %.5f (R^2 = %.2f)",
                        rampResult.kV, rampResult.kStatic, rampResult.rSquare));
            } else {
                telemetry.addLine(Misc.formatInvariant("kV = %.5f (R^2 = %.2f)",
                        rampResult.kStatic, rampResult.rSquare));
            }
            telemetry.addLine("Would you like to fit kA?");
            telemetry.addLine("Press (A) for yes, (B) for no");
            telemetry.update();

            state = State.ACCEL;
        }, drive);

        FunctionalCommand staticCommand = new FunctionalCommand(
                () -> {
                    telemetry.clearAll();
                    telemetry.addLine("Running...");
                    telemetry.update();

                    maxVel = rpmToVelocity(MAX_RPM);
                    finalVel = MAX_POWER * maxVel;
                    accel = (finalVel * finalVel) / (2.0 * DISTANCE);
                    rampTime = Math.sqrt(2.0 * DISTANCE / accel);

                    timeSamples = new ArrayList<>();
                    positionSamples = new ArrayList<>();
                    powerSamples = new ArrayList<>();

                    drive.setPoseEstimate(new Pose2d());

                    startTime = clock.seconds();
                },
                () -> {
                    elapsedTime = clock.seconds() - startTime;

                    double vel = accel * elapsedTime;
                    double power = vel / maxVel;

                    timeSamples.add(elapsedTime);
                    positionSamples.add(drive.getPoseEstimate().getX());
                    powerSamples.add(MAX_POWER);

                    drive.setDrivePower(new Pose2d(power, 0.0, 0.0));
                },
                staticFinishedCommand::schedule,
                () -> elapsedTime > rampTime,
                drive
        );

        InstantCommand accelFinishedCommand = new InstantCommand(() -> {
            drive.setDrivePower(new Pose2d(0.0, 0.0, 0.0));
            RegressionUtil.AccelResult accelResult = RegressionUtil.fitAccelData(
                    timeSamples, positionSamples, powerSamples, rampResult,
                    LoggingUtil.getLogFile(Misc.formatInvariant(
                            "DriveAccelRegression-%d.csv", System.currentTimeMillis())));

            telemetry.clearAll();
            telemetry.addLine("Constant power test complete");
            telemetry.addLine(Misc.formatInvariant("kA = %.5f (R^2 = %.2f)",
                    accelResult.kA, accelResult.rSquare));
            telemetry.update();

            reset(); // end the scheduler instance
        }, drive);

        FunctionalCommand accelCommand = new FunctionalCommand(
                () -> {
                    telemetry.clearAll();
                    telemetry.addLine("Running...");
                    telemetry.update();

                    maxPowerTime = DISTANCE / maxVel;

                    timeSamples.clear();
                    positionSamples.clear();
                    powerSamples.clear();

                    drive.setPoseEstimate(new Pose2d());
                    drive.setDrivePower(new Pose2d(MAX_POWER, 0.0, 0.0));

                    startTime = clock.seconds();
                },
                () -> {
                    elapsedTime = clock.seconds() - startTime;

                    timeSamples.add(elapsedTime);
                    positionSamples.add(drive.getPoseEstimate().getX());
                    powerSamples.add(MAX_POWER);
                },
                accelFinishedCommand::schedule,
                () -> elapsedTime > maxPowerTime,
                drive
        );

        aButton = new GamepadButton(gamepad, GamepadKeys.Button.A)
            .whenPressed(() -> {
                switch (state) {
                    case STATIC:
                        telemetry.clearAll();
                        telemetry.addLine(Misc.formatInvariant(
                                "Place your robot on the field with at least %.2f in of room in front", DISTANCE));
                        telemetry.addLine("Press (A) to begin");
                        telemetry.update();

                        fitIntercept = true;
                        staticCommand.schedule();
                        break;
                    case ACCEL:
                        telemetry.clearAll();
                        telemetry.addLine("Place the robot back in its starting position");
                        telemetry.addLine("Press (A) to continue");
                        telemetry.update();

                        state = State.ACCEL_STARTED;
                        break;
                    case ACCEL_STARTED:
                        accelCommand.schedule();
                        break;
                }
            });

        bButton = new GamepadButton(gamepad, GamepadKeys.Button.B)
            .whenPressed(() -> {
                switch (state) {
                    case STATIC:
                        telemetry.clearAll();
                        telemetry.addLine(Misc.formatInvariant(
                                "Place your robot on the field with at least %.2f in of room in front", DISTANCE));
                        telemetry.addLine("Press (A) to begin");
                        telemetry.update();

                        staticCommand.schedule();
                        break;
                    case ACCEL:
                        reset(); // end the scheduled commands
                        break;
                }
            });
    }

}

File: TeamCode/src/main/kotlin/org/firstinspires/ftc/teamcode/drive/opmode/AutomaticFeedforwardTuner.kt
package org.firstinspires.ftc.teamcode.drive.opmode

import org.firstinspires.ftc.teamcode.config.DriveConstants.MAX_RPM
import org.firstinspires.ftc.teamcode.config.DriveConstants.RUN_USING_ENCODER
import org.firstinspires.ftc.teamcode.config.DriveConstants.rpmToVelocity
import com.acmerobotics.dashboard.FtcDashboard
import com.acmerobotics.dashboard.config.Config
import com.acmerobotics.dashboard.telemetry.MultipleTelemetry
import com.acmerobotics.roadrunner.geometry.Pose2d
import com.acmerobotics.roadrunner.util.NanoClock
import com.qualcomm.robotcore.eventloop.opmode.Autonomous
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import com.qualcomm.robotcore.util.RobotLog
import org.firstinspires.ftc.robotcore.external.Telemetry
import org.firstinspires.ftc.robotcore.internal.system.Misc
import org.firstinspires.ftc.teamcode.config.CDConfig
import org.firstinspires.ftc.teamcode.drive.CDMecanumDrive
import org.firstinspires.ftc.teamcode.hardware.HardwareManager
import org.firstinspires.ftc.teamcode.util.LoggingUtil
import org.firstinspires.ftc.teamcode.util.RegressionUtil
import kotlin.math.sqrt

/*
 * Op mode for computing kV, kStatic, and kA from various drive routines. For the curious, here's an
 * outline of the procedure:
 *   1. Slowly ramp the motor power and record encoder values along the way.
 *   2. Run a linear regression on the encoder velocity vs. motor power plot to obtain a slope (kV)
 *      and an optional intercept (kStatic).
 *   3. Accelerate the robot (apply constant power) and record the encoder counts.
 *   4. Adjust the encoder data based on the velocity tuning data and find kA with another linear
 *      regression.
 */
@Config
@Autonomous(group = "drive")
class AutomaticFeedforwardTuner : LinearOpMode() {
    @kotlin.Throws(java.lang.InterruptedException::class)
    override fun runOpMode() {
        if (RUN_USING_ENCODER) {
            RobotLog.setGlobalErrorMsg(
                "Feedforward constants usually don't need to be tuned " +
                        "when using the built-in drive motor velocity PID."
            )
        }

        val telemetry: Telemetry = MultipleTelemetry(this.telemetry, FtcDashboard.getInstance().telemetry)

        val drive = CDMecanumDrive(HardwareManager(CDConfig(), hardwareMap))

        val clock: NanoClock = NanoClock.system()

        telemetry.addLine("Press play to begin the feedforward tuning routine")
        telemetry.update()

        waitForStart()

        if (isStopRequested) return

        telemetry.clearAll()
        telemetry.addLine("Would you like to fit kStatic?")
        telemetry.addLine("Press (Y/Δ) for yes, (B/O) for no")
        telemetry.update()

        var fitIntercept = false
        while (!isStopRequested) {
            if (gamepad1.y) {
                fitIntercept = true
                while (!isStopRequested && gamepad1.y) {
                    idle()
                }
                break
            } else if (gamepad1.b) {
                while (!isStopRequested && gamepad1.b) {
                    idle()
                }
                break
            }
            idle()
        }

        telemetry.clearAll()
        telemetry.addLine(
            Misc.formatInvariant(
                "Place your robot on the field with at least %.2f in of room in front", DISTANCE
            )
        )
        telemetry.addLine("Press (Y/Δ) to begin")
        telemetry.update()

        while (!isStopRequested && !gamepad1.y) {
            idle()
        }
        while (!isStopRequested && gamepad1.y) {
            idle()
        }

        telemetry.clearAll()
        telemetry.addLine("Running...")
        telemetry.update()

        val maxVel: Double = rpmToVelocity(MAX_RPM)
        val finalVel = MAX_POWER * maxVel
        val accel = (finalVel * finalVel) / (2.0 * DISTANCE)
        val rampTime: Double = sqrt(2.0 * DISTANCE / accel)

        val timeSamples: MutableList<Double> = java.util.ArrayList<Double>()
        val positionSamples: MutableList<Double> = java.util.ArrayList<Double>()
        val powerSamples: MutableList<Double> = java.util.ArrayList<Double>()

        drive.poseEstimate = Pose2d()

        var startTime: Double = clock.seconds()
        while (!isStopRequested) {
            val elapsedTime: Double = clock.seconds() - startTime
            if (elapsedTime > rampTime) {
                break
            }
            val vel = accel * elapsedTime
            val power = vel / maxVel

            timeSamples.add(elapsedTime)
            positionSamples.add(drive.poseEstimate.x)
            powerSamples.add(power)

            drive.setDrivePower(Pose2d(power, 0.0, 0.0))
            drive.updatePoseEstimate()
        }
        drive.setDrivePower(Pose2d(0.0, 0.0, 0.0))

        val rampResult: RegressionUtil.RampResult = RegressionUtil.fitRampData(
            timeSamples, positionSamples, powerSamples, fitIntercept,
            LoggingUtil.getLogFile(
                Misc.formatInvariant(
                    "DriveRampRegression-%d.csv", System.currentTimeMillis()
                )
            )
        )

        telemetry.clearAll()
        telemetry.addLine("Quasi-static ramp up test complete")
        if (fitIntercept) {
            telemetry.addLine(
                Misc.formatInvariant(
                    "kV = %.5f, kStatic = %.5f (R^2 = %.2f)",
                    rampResult.kV, rampResult.kStatic, rampResult.rSquare
                )
            )
        } else {
            telemetry.addLine(
                Misc.formatInvariant(
                    "kV = %.5f (R^2 = %.2f)",
                    rampResult.kStatic, rampResult.rSquare
                )
            )
        }
        telemetry.addLine("Would you like to fit kA?")
        telemetry.addLine("Press (Y/Δ) for yes, (B/O) for no")
        telemetry.update()

        var fitAccelFF = false
        while (!isStopRequested) {
            if (gamepad1.y) {
                fitAccelFF = true
                while (!isStopRequested && gamepad1.y) {
                    idle()
                }
                break
            } else if (gamepad1.b) {
                while (!isStopRequested && gamepad1.b) {
                    idle()
                }
                break
            }
            idle()
        }

        if (fitAccelFF) {
            telemetry.clearAll()
            telemetry.addLine("Place the robot back in its starting position")
            telemetry.addLine("Press (Y/Δ) to continue")
            telemetry.update()

            while (!isStopRequested && !gamepad1.y) {
                idle()
            }
            while (!isStopRequested && gamepad1.y) {
                idle()
            }

            telemetry.clearAll()
            telemetry.addLine("Running...")
            telemetry.update()

            val maxPowerTime = DISTANCE / maxVel

            timeSamples.clear()
            positionSamples.clear()
            powerSamples.clear()

            drive.poseEstimate = Pose2d()
            drive.setDrivePower(Pose2d(MAX_POWER, 0.0, 0.0))

            startTime = clock.seconds()
            while (!isStopRequested) {
                val elapsedTime: Double = clock.seconds() - startTime
                if (elapsedTime > maxPowerTime) {
                    break
                }

                timeSamples.add(elapsedTime)
                positionSamples.add(drive.poseEstimate.x)
                powerSamples.add(MAX_POWER)

                drive.updatePoseEstimate()
            }
            drive.setDrivePower(Pose2d(0.0, 0.0, 0.0))

            val accelResult: RegressionUtil.AccelResult = RegressionUtil.fitAccelData(
                timeSamples, positionSamples, powerSamples, rampResult,
                LoggingUtil.getLogFile(
                    Misc.formatInvariant(
                        "DriveAccelRegression-%d.csv", System.currentTimeMillis()
                    )
                )
            )

            telemetry.clearAll()
            telemetry.addLine("Constant power test complete")
            telemetry.addLine(
                Misc.formatInvariant(
                    "kA = %.5f (R^2 = %.2f)",
                    accelResult.kA, accelResult.rSquare
                )
            )
            telemetry.update()
        }

        while (!isStopRequested) {
            idle()
        }
    }

    companion object {
        var MAX_POWER: Double = 0.7
        var DISTANCE: Double = 100.0 // in
    }
}